core/server/src/test/java/alluxio/master/file/ConcurrentFileSystemMasterTest.java

```diff
@@ -0,0 +1,548 @@
+      if (i % 2 == 0) {
+        srcs[i] = dir1.join("file" + i);
+        dsts[i] = dir2.join("renamed" + i);
+      } else {
+        srcs[i] = dir2.join("file" + i);
+        dsts[i] = dir1.join("renamed" + i);
+      }
+      mFileSystemMaster.createFile(srcs[i], sCreatePersistedFileOptions);
+    }
+
+    int errors = concurrentRename(srcs, dsts);
+
+    // We should get no errors.
+    Assert.assertEquals(0, errors);
+
+    List<FileInfo> dir1Files =
+        mFileSystemMaster.listStatus(dir1, ListStatusOptions.defaults());
+    List<FileInfo> dir2Files =
+        mFileSystemMaster.listStatus(dir2, ListStatusOptions.defaults());
+
+    Assert.assertEquals(numThreads / 2, dir1Files.size());
+    Assert.assertEquals(numThreads / 2, dir2Files.size());
+
+    Collections.sort(dir1Files, new IntegerSuffixedPathComparator());
+    for (int i = 1; i < numThreads; i += 2) {
+      Assert.assertEquals(dsts[i].getName(), dir1Files.get(i / 2).getName());
+    }
+
+    Collections.sort(dir2Files, new IntegerSuffixedPathComparator());
+    for (int i = 0; i < numThreads; i += 2) {
+      Assert.assertEquals(dsts[i].getName(), dir2Files.get(i / 2).getName());
+    }
+  }
+
+  /**
+   * Helper for renaming a list of paths concurrently. Assumes the srcs are already created and
+   * dsts do not exist. Enforces that the run time of this method is not greater than twice the
+   * sleep time (to infer concurrent operations).
+   *
+   * @param src list of source paths
+   * @param dst list of destination paths
+   * @return how many errors occurred
+   */
+  private int concurrentRename(final AlluxioURI[] src, final AlluxioURI[] dst)
+      throws Exception {
+    final int numFiles = src.length;
+    final CyclicBarrier barrier = new CyclicBarrier(numFiles);
+    List<Thread> threads = new ArrayList<>(numFiles);
+    // If there are exceptions, we will store them here.
+    final ConcurrentHashSet<Throwable> errors = new ConcurrentHashSet<>();
+    Thread.UncaughtExceptionHandler exceptionHandler = new Thread.UncaughtExceptionHandler() {
+      public void uncaughtException(Thread th, Throwable ex) {
+        errors.add(ex);
+      }
+    };
+    for (int i = 0; i < numFiles; i++) {
+      final int iteration = i;
+      Thread t = new Thread(new Runnable() {
+        @Override
+        public void run() {
+          try {
+            AuthenticatedClientUser.set(TEST_USER);
+            barrier.await();
+            mFileSystemMaster.rename(src[iteration], dst[iteration]);
+          } catch (Exception e) {
+            Throwables.propagate(e);
+          }
+        }
+      });
+      t.setUncaughtExceptionHandler(exceptionHandler);
+      threads.add(t);
+    }
+    Collections.shuffle(threads);
+    long startNs = System.nanoTime();
+    for (Thread t : threads) {
+      t.start();
+    }
+    for (Thread t : threads) {
+      t.join();
+    }
+    long durationMs = (System.nanoTime() - startNs) / Constants.SECOND_NANO * Constants.SECOND_MS;
```

Comment:
durationMs will be zero as long as the renames took less than 1 second due to integer division by `Constants.SECOND_NANO`. Could we keep the duration units in `ms` at all times and use `System.currentTimeMillis()`?